crates/key-value-spin/src/store.rs

```diff
@@ -156,6 +157,140 @@ impl Store for SqliteStore {
                 .collect()
         })
     }
+
+    async fn get_many(&self, keys: Vec<String>) -> Result<Vec<Option<(String, Vec<u8>)>>, Error> {
+        task::block_in_place(|| {
+            let sql_value_keys: Vec<rusqlite::types::Value> =
+                keys.into_iter().map(rusqlite::types::Value::from).collect();
+            let ptr = Rc::new(sql_value_keys);
+            let row_iter: Vec<Result<(String, Vec<u8>), Error>> = self.connection
+                .lock()
+                .unwrap()
+                .prepare_cached("SELECT key, value FROM spin_key_value WHERE store=:name AND key IN rarray(:keys)")
+                .map_err(log_error)?
+                .query_map((":name", &self.name, ":keys", ptr), |row| {
+                    <(String, Vec<u8>)>::try_from(row)
+                })
+                .map_err(log_error)?
+                .map(|r: Result<(String, Vec<u8>), rusqlite::Error>| r.map_err(log_error))
+                .collect();
+
+            let mut keys_and_values: Vec<Option<(String, Vec<u8>)>> = Vec::new();
+            for row in row_iter {
+                let res = row.map_err(log_error)?;
+                keys_and_values.push(Some(res));
+            }
+            Ok(keys_and_values)
+        })
+    }
+
+    async fn set_many(&self, key_values: Vec<(String, Vec<u8>)>) -> Result<(), Error> {
+        task::block_in_place(|| {
+            let mut binding = self.connection.lock().unwrap();
+            let tx = binding.transaction().map_err(log_error)?;
+            for kv in key_values {
+                tx.prepare_cached(
+                    "INSERT INTO spin_key_value (store, key, value) VALUES ($1, $2, $3)
+                     ON CONFLICT(store, key) DO UPDATE SET value=$3",
+                )
+                .map_err(log_error)?
+                .execute(rusqlite::params![&self.name, kv.0, kv.1])
+                .map_err(log_error)
+                .map(drop)?;
+            }
+            tx.commit().map_err(log_error)
+        })
+    }
+
+    async fn delete_many(&self, keys: Vec<String>) -> Result<(), Error> {
+        task::block_in_place(|| {
+            let sql_value_keys: Vec<rusqlite::types::Value> =
+                keys.into_iter().map(rusqlite::types::Value::from).collect();
+            let ptr = Rc::new(sql_value_keys);
+            self.connection
+                .lock()
+                .unwrap()
+                .prepare_cached("DELETE FROM spin_key_value WHERE store=:name AND key IN (:keys)")
+                .map_err(log_error)?
+                .execute((":name", &self.name, ":keys", ptr))
+                .map_err(log_error)
+                .map(drop)
+        })
+    }
+
+    // The assumption with increment is that if the value for the key does not exist, it will be
+    // assumed to be zero. In the case that we are unable to unmarshal the value into an i64 an error will be returned.
+    async fn increment(&self, key: String, delta: i64) -> Result<i64, Error> {
```

Comment:
This is an interesting design niggle - we have not previously defined an encoding for integers, and I don't think `wasi-keyvalue` defines one either.  All the "set" APIs are byte arrays.  So how do I set an initial value to which I can later apply `increment`?  Hopefully this won't be an issue in practice, but might be worth considering in the wasi-keyvalue specification.

Reply:
Nice catch! I'm just going with little endian encoding of i64s. I think the guidance on using increment right now is to start with a key/value that doesn't exist and "increment" it to your value of choice. Then use increment or get from there on out.

I do believe this is something that we'll need to consider for future interface revisions.